### DISCUSSION TITLE
Dell Z440/Z240 hangs on "iPXE initializing devices..."

### DISCUSSION THREAD

**Discussion by hackdroot** [ipxe/ipxe#418] [Q&A]
We recently updated ipxe from 1.0.0 to 1.21.1 to resolve issues with newer hardware, but are now having issues with hardware that previously worked (using UEFI), specifically Dell Z440/Z240 with i218lm
 
8086:15a0, driver i218lm-2

Currently using DEBUG="device,pci,init,intel,rbg,drbg"

Debug stops at: `INTEL 0xc3409cd0 link status is 00080483`

I have jpeg screenshot images, but cant seem to attach them successfully.

**Comment by ccy** [comment]
Can you try using `http://boot.ipxe.org/snponly.efi` to boot from PXE?

**Reply by hackdroot** [reply]
I think that may have resolved the issue. I'll have to test against our other hardware and ensure it works across the board. Thanks!

**Reply by ccy** [reply]
Can you describe your network switch model? Do you setup jumbo frame?

**Comment by NiKiZe** [comment]
You have one local build working, and latest failing correct? if so try [git bisect](https://ipxe.org/howto/bisect) to locate where this issue started. (#427 has similarities, but I understood that as failing to run iPXE at all, but maybe that is the difference in debug)

**Comment by cyrozap** [comment]
I have an HP Z440, with the same NIC, failing the same way when running [Arch Linux's lightly-patched version of 1.21.1](https://gitlab.archlinux.org/archlinux/packaging/packages/ipxe).

After making a few more patches to fix the build and building with `DEBUG="device,pci,init,intel,rbg,drbg"`, I get the following messages:

```
>>Start PXE over IPv4.
  Station IP address is <removed>

  Server IP address is <removed>
  NBP filename is ipxe-arch.efi
  NBP filesize is 1037504 Bytes
 Downloading NBP file...

  Succeed to download NBP file.
iPXE Initalising devices...INIT startup efi_tick...
INIT startup devices...
Adding 3c509 root bus
Adding EFI root bus
0000:00:19.0 (8086:15a0) has driver "i218lm-2"
0000:00:19.0 has mem fb100000 io 3040 irq 255
0000:00:19.0 Latency timer is unreasonably low at 0. Setting to 32.
INTEL 0xd41ced70 forced MAC reset (18180260/40080483 was 18180240/40080083)
INTEL 0xd41ced70 has autoloaded MAC address <removed>
INTEL 0xd41ced70 link status is 40080483
```

Is there any quick change I could make to get iPXE to fall back to the UEFI network driver or something? I'm also willing to do some more debugging and actually fix the issue, but I'm not sure where to start with that.

**Reply by NiKiZe** [reply]
If you want an EFI build without native drivers, use the snp.efi target.

You mentioned that you "have made patches to fix the build" what does that mean? Please use current git master and report any build issues you might have.

You are probably here seeing a working Intel initialization, and then iPXE hangs on other driver, maybe even on SNP.

**Reply by cyrozap** [reply]
> You mentioned that you "have made patches to fix the build" what does that mean? Please use current git master and report any build issues you might have.

I had to apply #1457 and backport 6ca597eee9f95b846a3c2dc1231e63cfc02272c1 to get the build to stop failing, that's all.

> You are probably here seeing a working Intel initialization, and then iPXE hangs on other driver, maybe even on SNP.

Looks like that's exactly what was happening. After enabling a bunch of `DEBUG` options (see bottom of this comment for the exact string, since it's rather long), I saw that it was getting stuck on the following lines:

```
...
SNPDEV 0xd412ffe8 installed for net0 as device PciRoot(0x0)/Pci(0x19,0x0)/MAC(<removed>,0x1)
Settings 0xd4130658 ("net0.link") registered
IPv6 net0 has address fe80::<removed>/64
PRIVKEY has no private key
INTEL 0xd412fe90 link status is 40080483
NETDEV net0 link is up
SNPDEV 0xd412ffe8 link is up
EFIPCI 0000:00:19.0 using driver "i218lm-2"
EFIDRV PciRoot(0x0)/Pci(0x19,0x0) using driver "PCI"
SNPDEV 0xd412ffe8 START
SNPDEV 0xd412ffe8 allowing start while claimed
SNPDEV 0xd412ffe8 INITIALIZE (0 extra RX, 0 extra TX)
SNPDEV 0xd412ffe8 ignoring initialization while claimed
SNPDEV 0xd412ffe8 RECEIVE_FILTERS 00000005&~00000002 reset 0 mcast
SNPDEV 0xd412ffe8 RECEIVE_FILTERS 0000000d&~0000000a 0 mcast
SNPDEV 0xd412ffe8 RECEIVE_FILTERS 0000000d&~0000000a 0 mcast
SNPDEV 0xd412ffe8 RECEIVE_FILTERS 0000000d&~0000000a 0 mcast
SNPDEV 0xd412ffe8 RECEIVE_FILTERS 0000000d&~0000000a 0 mcast
EFIDRV PciRoot(0x0)/Pci(0x19,0x0)/MAC(<removed>,0x1)/IPv6(0000:0000:0000:0000:0000:0000:0000:0000) has driver "SNP"
EFIDRV PciRoot(0x0)/Pci(0x19,0x0)/MAC(<removed>,0x1)/IPv6(0000:0000:0000:0000:0000:0000:0000:0000) DRIVER_START
EFIDRV PciRoot(0x0)/Pci(0x19,0x0)/MAC(<removed>,0x1)/IPv6(0000:0000:0000:0000:0000:0000:0000:0000) is not supported by driver "PCI": Error 0x7f10e183 (http://ipxe.org/7f10e183)
EFIDRV PciRoot(0x0)/Pci(0x19,0x0)/MAC(<removed>,0x1)/IPv6(0000:0000:0000:0000:0000:0000:0000:0000) is not supported by driver "NII": Error 0x7f684083 (http://ipxe.org/7f684083)
```

I don't know iPXE's code very well, but it seemed to me the SNP driver was trying to access the IPv6 network interface, despite having booted the system in IPv4 PXE mode, and I assume that this was what was causing iPXE to get stuck.

I may try building from git master later to see if that issue still exists. For now, though, I was able to work around the issue by changing the package from building `bin-x86_64-efi/ipxe.efi` to building `bin-x86_64-efi/intel--snponly.efi`. With only those drivers in the image, I was able to successfully boot it over the network.

Thanks for your help!

<hr>

The `DEBUG` options I used:

```
DEBUG="acm,acpi,acpipwr,acpi_settings,acpi_timer,aes,ansicol,ansicoldef,aoe,arbel,arp,asn1,ata,axge,bios_console,bitmap,blocktrans,bzimage,cachedhcp,certstore,cms,com32,com32_call,comboot,cpio,cpuid,cpuid_settings,debug,deflate,der,dhcp,dhcpopts,dhcpv6,dm96xx,dns,dns_test,drbg,dummy_sanboot,eapol,ecm,efi_autoboot,efi_block,efi_bofm,efi_driver,efi_entropy,efi_fbcon,efi_fdt,efi_file,efi_hii,efi_image,efi_init,efi_local,efi_path,efi_pci,efi_pxe,efi_snp,efi_snp_hii,efi_time,efi_timer,efi_usb,efi_utils,efi_veto,efi_watchdog,efi_wrap,efx_common,efx_hunt,ehci,elfboot,elf,ena,entropy,eoib,eth_slow,exanic,fbcon,fc,fcels,fcns,fcoe,fcp,fdt,flexboot_nodnic,fragment,ftp,gdbmach,golan,guestinfo,guestrpc,hash_df,hermon,hmac_drbg,httpauth,httpbasic,httpblock,httpconn,httpcore,httpdigest,httpntlm,hvm,hyperv,i2c_bit,ib_cm,ib_cmrc,ib_mcast,ib_mi,ib_packet,ib_pathrec,ib_sma,ib_smc,ib_srp,icmp,icmpv6,icplus,image,infiniband,init,initrd,int13,intel,intelvf,intelx,intelxl,intelxlvf,intelxvf,interface,iobuf_test,iphone,ipoib,ipv4,ipv6,iscsi,job,lan78xx,librm_mgmt,linda,linebuf_test,malloc,md4,md5,memmap,memmap_settings,mii,multiboot,myson,natsemi,nbi,ncm,ndp,neighbour,net80211,netdevice,netdev_settings,netfront,netvsc,nfs_open,nii,ntlm,ntlm_test,ntp,nvo,nvsvpd,ocsp,ocsp_test,open,params,parseopt,pccrc,pccrd,pcibackup,pci,pciea,pcimsix,pcivpd,peerblk,peerdisc,peerdist,peermux,pem,pending,phantom,ping,pinger,pixbuf,png,pnm,privkey,process,profile,profile_test,pxe_call,pxe_image,pxe_preboot,pxe_undi,qib7322,rbg,rc80211,rdtsc_timer,readline,realtek,refcnt,relocate,resolv,retry,rhine,rndis,rootcert,rsa,rsdp,rtc_entropy,rtc_time,runtime,sanboot,script,scsi,sdi,segment,serial,setjmp_test,settings,settings_test,sfc_hunt,sha1,sha256,sha512,skeleton,slam,smbios,smbios_settings,smsc75xx,smsc95xx,smscusb,snpnet,snponly,spi_bit,srp,stp,tcp,tftp,tg3,tg3_hw,tg3_phy,threewire,thunderx,time,timer,tls,uart,udp,uhci,uri,usbblk,usb,usbhid,usbhub,usbio,usbkbd,usbnet,validator,velocity,vesafb,virtio-net,vlan,vmbus,vmware,vmxnet3,wep,wpa,wpa_ccmp,wpa_psk,wpa_tkip,x509,x86_uart,xenbus,xengrant,xenstore,xferbuf,xfer,xhci,xsigo"
```

**Reply by NiKiZe** [reply]
So why are you not using git master to begin with?
We always expect latest version (git master) to be used before we do any kind of attempts to debug issues.

**Reply by cyrozap** [reply]
Arch builds iPXE from the latest stable version, which is version 1.21.1 (though it looks like there's [an open MR to update it](https://gitlab.archlinux.org/archlinux/packaging/packages/ipxe/-/merge_requests/2)). I was using Arch's version of iPXE because at the time I selected it, I was unsure how to get the current version of iPXE to properly check the image signatures and mirror HTTPS certs, and due to #606 I was unable to use the mainline version of iPXE to boot from some HTTPS mirrors.

It appears all these issues have since been fixed in git master.

After cloning the iPXE master branch and copying the configuration and certificates from the Arch iPXE package's repo, I ran the following command:

```bash
make EMBED="arch.ipxe" CERT="isrgrootx1.pem,lets-encrypt-r3.pem,codesigning_pierre_archlinux.pem" TRUST="isrgrootx1.pem,lets-encrypt-r3.pem,codesigning_pierre_archlinux.pem" NO_WERROR=1 bin-x86_64-efi/ipxe.efi
```

This successfully built iPXE with no errors, and all my systems--including the Z440--were able to boot Arch over the internet. They were also able to use the HTTPS mirror that was giving me trouble before.
